### PR TITLE
Implement umask for Windows

### PR DESCRIPTION
**Pull request by adferrand**
This PR gets its root from an observation I did on current version of Certbot (1.3.0): the `renewal-hooks` directory in Certbot configuration directory is created on Windows with write permissions to everybody.

I thought it was a critical bug since this directory contains hooks that are executed by Certbot, and you certainly do not want this folder to be open to any malicious hook that could be inserted by everyone, then executed with administrator privileges by Certbot.

Turns out for this specific problem that the bug is not critical for the hooks, because the scripts are expected to be in subdirectories of `renewal-hooks` (namely `pre`, `post` and `deploy`), and these subdirectories have proper permissions because we set them explicitly when Certbot is starting.

Still, there is a divergence here between Linux and Windows: on Linux all Certbot directories without explicit permissions have at maximum `0o755` permissions by default, while on Windows it is a `0o777` equivalent. It is not an immediate security risk, but it is definitly error-prone, not expected, and so a potential breach in the future if we forget about it.

Root cause is that umask is not existing in Windows. Indeed under Linux the umask defines the default permissions when you create a file or a directory. Python takes that into account, with an API for `os.open` and `os.mkdir` that expose a `mode` parameter with default value of `0o777`. In practice it is never `0o777` (either you the the `mode` explictly or left the default one) because the effective mode is masked by the current umask value in the system: on Linux it is `0o022`, so files/directories have a maximum mode of `0o755` if you did not set the umask explicitly, and it is what it is observed for Certbot.

However on Windows, the `mode` value passed (and got from default) to the `open` and `mkdir` of `certbot.compat.filesystem` module is taken verbatim, since umask does not exit, and then is used to calculate the DACL of the newly created file/directory. So if the mode is not set explicitly, we end up with files and directories with `0o777` permissions.

This PR fixes this problem by implementing a umask behavior in the `certbot.compat.filesystem` module, that will be applied to any file or directory created by Certbot since we forbid to use the `os` module directly.

The implementation is quite straight-forward. For Linux the behavior is not changed. On Windows a `mask` parameter is added to the function that calculates the DACL, to be invoked appropriately when file or directory are created. The actual value of the mask is taken from an internal class of the `filesystem` module: its default value is `0o755` to match default umasks on Linux, and can be changed with the new method `umask` that have the same behavior than the original `os.umask`. Of course `os.umask` becomes a forbidden function and `filesystem.umask` must be used instead.

Existing code that is impacted have been updated, and new unit tests are created for this new function.

### CODE REVIEW

**Comment by adferrand** [comment]
@bmw, I block on answering the comments here until #8010 is merged, because it will require some adaptations.

File: certbot/certbot/compat/filesystem.py [hunk 1]
```diff
@@ -21,6 +21,27 @@
     POSIX_MODE = False
 
 
+# Windows umask implementation, since Windows does not have a concept of umask by default.
+# We choose 022 as initial value since it is the default one on Linux, and it is a decent
+# choice to not have write permissions for group owner and everybody by default.
+class WindowsUmask:
+    """Store the current umask to apply on Windows"""
+    def __init__(self):
+        self._mask = 0o022
```
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot/certbot/compat/filesystem.py:30]: Indeed, simple is better than complex.

File: certbot/certbot/compat/filesystem.py [hunk 2]
```diff
@@ -21,6 +21,27 @@
     POSIX_MODE = False
 
 
+# Windows umask implementation, since Windows does not have a concept of umask by default.
+# We choose 022 as initial value since it is the default one on Linux, and it is a decent
```
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot/certbot/compat/filesystem.py:25]: Fixed.

File: certbot/certbot/compat/filesystem.py [hunk 3]
```diff
@@ -21,6 +21,27 @@
...
+# choice to not have write permissions for group owner and everybody by default.
+class WindowsUmask:
```
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot/certbot/compat/filesystem.py:27]: Done.

File: certbot/certbot/compat/filesystem.py [hunk 4]
```diff
@@ -42,6 +63,23 @@ def chmod(file_path, mode):
         _apply_win_mode(file_path, mode)
 
 
+def umask(mask):
+    # type: (int) -> int
+    """
+    Set the current numeric umask and return the previous umask. On Linux, the built-in umask
+    method is used. On Windows, our Certbot-side implementation is used.
+    :param int mask: The user file-creation mode mask to apply.
```
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot/certbot/compat/filesystem.py:71]: Fixed. And for the other functions also.

File: certbot/tests/compat/filesystem_test.py [hunk 5]
```diff
@@ -150,6 +150,57 @@ def test_user_admin_dacl_consistency(self):
         self.assertEqual(security_dacl.GetSecurityDescriptorDacl().GetAceCount(), 2)
 
 
+class UmaskTest(TempDirTestCase):
+    def test_umask_on_dir(self):
+        previous_umask = filesystem.umask(0o022)
+
```
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot/tests/compat/filesystem_test.py:157]: Done.

File: certbot/tests/compat/filesystem_test.py [hunk 6]
```diff
@@ -150,6 +150,57 @@ def test_user_admin_dacl_consistency(self):
...
+            dir3 = os.path.join(self.tempdir, 'dir3')
+            filesystem.mkdir(dir3, mode=0o777)
+            self.assertTrue(filesystem.check_mode(dir3, 0o700))
+        finally:
+            filesystem.umask(previous_umask)
+
+    def test_umask_on_file(self):
+        previous_umask = filesystem.umask(0o022)
+
```
[the remark this answers is not in the document]
Comment (adferrand) [reply certbot/tests/compat/filesystem_test.py:177]: Done.

File: certbot-apache/setup.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
 # acme/certbot version.
 install_requires = [
     'acme>=0.29.0',
-    'certbot>=1.1.0',
+    'certbot>=1.5.0.dev0',
```
Comment (bmw) [inline certbot-apache/setup.py:15]:
```suggestion
    'certbot>=1.6.0.dev0',
```

File: certbot/certbot/compat/filesystem.py [hunk 8]
```diff
@@ -232,39 +270,38 @@ def makedirs(file_path, mode=0o777):
...
-        # leaf) created with the given mode.
-        current_umask = os.umask(0)
+    current_umask = umask(0)
+    try:
+        if POSIX_MODE:
+            # Since Python 3.7, os.makedirs does not set the given mode to the intermediate
+            # directories that could be created in the process. To keep things safe and consistent
+            # on all Python versions, we set the umask accordingly to have all directories
+            # (intermediate and leaf) created with the given mode.
+            umask(current_umask | 0o777 ^ mode)
```
Comment (bmw) [inline certbot/certbot/compat/filesystem.py:285]: I think we need to do something similar to this in the Windows case.

Comment (adferrand) [reply certbot/certbot/compat/filesystem.py]: It was the intention indeed. Done.

**Comment by adferrand** [comment]
Yes it was almost done, I was on the negative testing step. PR is ready and corrections are applied, including changelog entries.

File: certbot/CHANGELOG.md [hunk 9]
```diff
@@ -7,10 +7,15 @@ Certbot adheres to [Semantic Versioning](https://semver.org/).
 ### Added
 
 * Add minimal code to run Nginx plugin on NetBSD.
+* Function `certbot.compat.filesystem.umask` is a drop-in replacement for `os.umask`
+  implementing umask for both UNIX and Windows systems.
 
 ### Changed
 
 * Allow session tickets to be disabled in Apache when mod_ssl is statically linked.
+* Certbot behaves similarly on Windows than on UNIX systems regarding umask, and
```
Comment (bmw) [inline certbot/CHANGELOG.md:16]:
```suggestion
* Certbot behaves similarly on Windows to on UNIX systems regarding umask, and
```

**Pull request merged**